topdown/tokens.go

```diff
@@ -864,8 +864,90 @@ func builtinJWTDecodeVerify(a ast.Value, b ast.Value) (v ast.Value, err error) {
 	return arr, nil
 }
 
+func extractUnverifiedPayloadAsAST(token ast.Value) (ast.Value, error) {
+	jwtToken, err := decodeJWT(token)
+	if err != nil {
+		return nil, err
+	}
+
+	p, err := builtinBase64UrlDecode(ast.String(jwtToken.payload))
+	if err != nil {
+		return nil, fmt.Errorf("JWT payload had invalid encoding: %v", err)
+	}
+
+	payload, err := extractJSONObject(string(p.(ast.String)));
+	if err != nil {
+		return nil, err
+	}
+
+	return payload, nil
+}
+
+// Implements full JWT decoding, validation and verification.
+func builtinOpenIdConnectTokenVerifyAndParse(a ast.Value, b ast.Value) (v ast.Value, err error) {
```

Comment:
We may not need it right away, but it might be good to switch to the newer builtin signature https://github.com/open-policy-agent/opa/blob/master/topdown/builtins.go#L50 so that we have access to the builtin context and the cache provided there.

The advantage is that if for whatever reason someone had a policy that required re-evaluating the builtin function we could (for each query evaluation) avoid re-computing anything and return the same value immediately.

Reply:
Oh very nice. A decision caching layer. Thanks for the heads up on this.

Though, one followup question. How long does the context cache last? I ask because token validity is directly dependent on time, so a re-eval of the same inputs may result in a different decision. due to a token's validity period.  But if it's only for the same request, I would venture using the caching layer is far more a benefit than a risk.

Reply:
It would be on a per-query evaluation basis, so like using the OPA server that roughly means per-request to OPA.